File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/base/NEdgeInsets.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.base

import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asDouble
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asMap
import dev.note11.flutter_naver_map.flutter_naver_map.util.DisplayUtil

internal data class NEdgeInsets(
    private val left: Double,
    private val top: Double,
    private val right: Double,
    private val bottom: Double,
    val leftPx: Int,
    val topPx: Int,
    val rightPx: Int,
    val bottomPx: Int,
) {

    fun <T> use(func: (left: Int, top: Int, right: Int, bottom: Int) -> T): T =
        func.invoke(leftPx, topPx, rightPx, bottomPx)

    companion object {
        fun fromMessageable(args: Any): NEdgeInsets = args.asMap().let { map ->
            val left = map["left"]!!.asDouble()
            val top = map["top"]!!.asDouble()
            val right = map["right"]!!.asDouble()
            val bottom = map["bottom"]!!.asDouble()
            NEdgeInsets(
                left, top, right, bottom,
                leftPx = DisplayUtil.dpToPx(left),
                topPx = DisplayUtil.dpToPx(top),
                rightPx = DisplayUtil.dpToPx(right),
                bottomPx = DisplayUtil.dpToPx(bottom)
            )
        }
    }
}

File: ios/flutter_naver_map/Sources/flutter_naver_map/model/flutter_default_custom/NEdgeInsets.swift
import UIKit

internal struct NEdgeInsets {
    let left: CGFloat
    let top: CGFloat
    let right: CGFloat
    let bottom: CGFloat

    var uiEdgeInsets: UIEdgeInsets {
        UIEdgeInsets(top: top, left: left, bottom: bottom, right: right)
    }

    static func fromMessageable(_ args: Any) -> NEdgeInsets {
        let d = asDict(args, valueCaster: asCGFloat)
        return NEdgeInsets(
                left: d["left"]!,
                top: d["top"]!,
                right: d["right"]!,
                bottom: d["bottom"]!
        )
    }
}
